File: boj/2878.cpp
#include <bits/stdc++.h>
#define ll long long int
#define all(x) x.begin(), x.end()
#define MX 101010
#define INT __int128_t
using namespace std;

__int128_t pw(int k) {
    __int128_t ret = 1;
    __int128_t base = 2;
    while (k) {
        if (k & 1)
            ret *= base;
        base *= base;
        k>>=1;
    }
    return ret;
}

vector<INT> arr;
unsigned long long m, n, tot;
INT MOD = pw(64);
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    cin >> m >> n;
    int mx = -1;
    for (int i = 0; i < n; i++) {
        int t;
        cin >> t;
        arr.push_back(t);
        tot += arr[i];
    }
    if(tot < m){
        cout << 0 <<'\n';
        return 0;
    }
    tot -=m;
    sort(all(arr));
    INT answer = 0;
    for (int i = 0; i < n; i++) {
        INT tmp = min(arr[i], tot/(__int128_t)(n-i));
        answer += tmp * tmp;
        tot -= tmp;
    }
    unsigned long long ans = answer % MOD;
    cout << ans;
}

File: boj/2878.py
import sys
m, n = map(int, sys.stdin.readline().split())
data = []
tot = 0
for i in range(n):
    tmp = int(sys.stdin.readline())
    tot += tmp
    data.append(tmp)


if tot < m:
    sys.stdout.write('0')
    exit(0)

tot -= m

answer = 0
mod = (1 << 64)
r = tot // n
x = tot % n

for i in range(n):
    answer = (answer + (r + (x > 0))**2) % mod
    x -= 1


sys.stdout.write(str(answer))
